### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by worseTyler**

### CODE REVIEW

**Comment by NPThompson** [comment]
I will review this.

File: GenericCollectionClass/GenericCollectionClass.Tests/NumSetTests.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+            bool result = numSetOne != numSetTwo;
+
+            Assert.IsTrue(result);
+        }
+
+        [TestMethod]
+        public void NotEqualsOperator_GivenSameNumSet_ReturnFalse()
+        {
+            (NumSet numSetOne, NumSet numSetTwo) = this.CreateSameNumSet();
```
Comment (NPThompson) [inline GenericCollectionClass/GenericCollectionClass.Tests/NumSetTests.cs:122]: I like this style of initializing test elements in a tuple. It's very concise. Am going to steal it for future assignments.

File: GenericCollectionClass/GenericCollectionClass/NumSet.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        public override bool Equals(object? obj)
+        {
+            if (obj is not NumSet numSet)
+            {
+                return false;
+            }
+
+            return obj.GetHashCode() == this.GetHashCode();
```
Comment (NPThompson) [inline GenericCollectionClass/GenericCollectionClass/NumSet.cs:52]: I don't think this is correct. It's entirely possible( though very unlikely if your hash function is good enough and you're using large enough integers ) that two different sets could have the same Hash code. 
For this procedure to work correctly it needs to make sure that every element in 'this' is also in 'obj', and vice-versa.

File: GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            this.StreamWriter = new StreamWriter(filePath
+                ?? throw new ArgumentNullException($"{nameof(SetWriter)} cannot be instantiated because of a null filePath"));
+        }
+
+        private StreamWriter StreamWriter { get; }
+
+        public void WriteToStream(NumSet set)
+        {
+            this.StreamWriter.Write(set.ToString());
```
Comment (NPThompson) [inline GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs:20]: "The best solution to a problem is usually the easiest" ~ GlaDOS

**Review by NPThompson — COMMENTED** [review]
Brief:
- Both projects created with appropriate properties: ✔
- NumSet constructor accepts variable arguments: ✔
- NumSet ToString implemented correctly: ✔
- NumSet equality operators implemented correctly: ❌
- NumSet fully unit tested: ✔
- SetWriter created in nested namespace: ✔ 
- SeWriter writes set to file: ✔ 
- SetWriter properly implements IDisposable: ✔ 
- SetWriter fully unit tested: ✔ 

## Extra Credit
NumSet contains cast operator that is unit tested: ✔
Projects updated to properly use Directory.Build.props: ✔?

Looks Nice. 
NumSet.Equals() is flawed, tho: since it uses the hash code of the two sets to perform its comparison. Under the right circumstances, unequal sets will map to the same hash code, so NumSet.Equals() will (very rarely) treat unequal sets as equal. I left a comment on NumSet.Equals() to the same effect. The only correct way is to compare element by element and make sure all elements are in both sets.

A few nitpicky things: If the NumSet is going to behave like a set, it's Constructor should be reworked to prevent duplicate elements. Also, there is no test that makes sure that modifying the array derived from the NumSet doesn't modify the set contents.

File: GenericCollectionClass/GenericCollectionClass.Tests/NumSetTests.cs [hunk 4]
```diff
@@ -0,0 +1,163 @@
+namespace GenericCollectionClass.Tests
+{
+    using System;
+    using System.Collections.Generic;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass.Tests/NumSetTests.cs:7]: I think you may find using [ data rows ](https://docs.microsoft.com/en-us/dotnet/api/system.data.datarow?view=net-5.0) to be helpful in future assignments.

File: GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
+﻿namespace GenericCollectionClass.SetWriterSpace
+{
+    using System;
+    using System.IO;
+
+    public class SetWriter : IDisposable
+    {
+        private bool disposedValue;
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs:8]:
```suggestion
        private bool _DisposedValue;
```

File: GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+        public SetWriter(string filePath)
+        {
+            this.StreamWriter = new StreamWriter(filePath
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs:12]:
```suggestion  
            StreamWriter = new StreamWriter(filePath
```

File: GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs (hunk 3 above)
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs:20]:
```suggestion
            StreamWriter.Write(set.ToString());
```
You also may consider using `WriteLine` here.

File: GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!this.disposedValue)
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs:32]:
```suggestion
            if (!_DisposedValue)
```

File: GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            {
+                if (disposing)
+                {
+                    this.StreamWriter?.Close();
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs:36]:
```suggestion

```
You only need `Dispose` here.

File: GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                    this.StreamWriter?.Dispose();
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs:37]:
```suggestion
                    this.StreamWriter.Dispose();
```

File: GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
+﻿namespace GenericCollectionClass.SetWriterSpace
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass/SetWriterSpace/SetWriter.cs:1]: I see that you have a folder in your project for this namespace, just be aware that you do not _HAVE_ to have a folder to create a nested namespace.

File: GenericCollectionClass/GenericCollectionClass.Tests/SetWriterSpace/SetWriterTests.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            string filePath = Path.GetTempFileName();
+            try
+            {
+                using (SetWriter setWriter = new (filePath))
+                {
+                    Assert.IsNotNull(setWriter);
+                }
+            }
+            finally
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass.Tests/SetWriterSpace/SetWriterTests.cs:22]: 💯

File: GenericCollectionClass/GenericCollectionClass.Tests/SetWriterSpace/SetWriterTests.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            }
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void Create_GivenInvalidFilePath_ThrowException()
+        {
+#pragma warning disable SA1312 // Discarding, naming doesn't matter
+            SetWriter _ = new (null!);
+#pragma warning restore SA1312 // Discarding, naming doesn't matter
```
Comment (mmwoodfo) [inline GenericCollectionClass/GenericCollectionClass.Tests/SetWriterSpace/SetWriterTests.cs:55]: An alternative
```suggestion
            new SetWriter(null!);
```
or
```suggestion
            _ = new SetWriter(null!);
```

**Pull request closed without merging**